dev-docs/design/how-lit-html-works.md

```diff
@@ -0,0 +1,425 @@
+
+All that is left is to iterate over the previously created `Part`s and call `_$setValue` with the new values. For demonstrative purposes, inlined this looks like:
+
+```js
+AttributePart._$setValue('odd');
+ChildPart._$setValue(1);
+EventPart._$setValue(() => render(counterUi(2), container));
+```
+
+This results in the DOM updating with a green underlined number one, and the event listener is swapped with the new listener.
+
+# Future work
+
+lit-html is fast, but what if it could be faster!?
+
+- [Declarative syntax for creating DOM Parts](https://github.com/WICG/webcomponents/issues/990). This could remove the requirement for lit-html to walk the `<template>` element's tree and manually instantiate Parts. This reduces the tree walk done in the **Create** phase.
+- Precompiling lit-html `html` tag functions (tracked in https://github.com/lit/lit/issues/189). This would allow opting into doing the **Prepare** phase in a compile step done by build tooling. Runtime would no longer have a **Prepare** phase.
+
+# Appendix
+
+## Parts
+
+A `Part` is a lit-html concept and represents the location of an expression in the `html` tagged template literal:
+
+| Part                   | Description                                                        | Authored                               |
+| ---------------------- | ------------------------------------------------------------------ | -------------------------------------- |
+| `ChildPart`            | Expressions in HTML child position                                 | `` html`<div>${...}</div>`  ``         |
+| `AttributePart`        | Expressions in HTML attribute value position                       | `` html`<input id="${...}">`  ``       |
+| `BooleanAttributePart` | Expressions in a boolean attribute value (name prefixed with `?`)  | `` html`<input ?checked="${...}">`  `` |
+| `EventPart`            | Expressions in an event listener position (name prefixed with `@`) | `` html`<input @click=${...}>`  ``     |
+| `PropertyPart`         | Expressions in property value position (name prefixed with `.`)    | `` html`<input .value=${...}>`  ``     |
+| `ElementPart`          | Expressions on the element tag                                     | `` html`<input ${...}>`  ``            |
+
+In all the cases above the authored code pass an expression into `${<expr>}` which represents a dynamic binding to the template, and the different part types implement how the value is committed to the DOM. For instance the `EventPart` in `` html`<input @click=${() => console.log('clicked')}`  `` will take the user provided function, and manage `addEventListener` and `removeEventListener` calls on the DOM such that the passed function is called when the click event is triggered.
```

Comment:
Should it be this?
```suggestion
In all the cases above the authored code pass an expression into `${...}` which represents a dynamic binding to the template, and the different part types implement how the value is committed to the DOM. For instance the `EventPart` in `` html`<input @click=${() => console.log('clicked')}`  `` will take the user provided function, and manage `addEventListener` and `removeEventListener` calls on the DOM such that the passed function is called when the click event is triggered.
```
or we fix it to `${<expr>}` in the table.

Reply:
depending on whether you take the suggestion about changing the element for the event part:
```suggestion
In all the cases above the authored code pass an expression into `${<expr>}` which represents a dynamic binding to the template, and the different part types implement how the value is committed to the DOM. For instance the `EventPart` in `` html`<button @click=${() => console.log('clicked')}`  `` will take the user provided function, and manage `addEventListener` and `removeEventListener` calls on the DOM such that the passed function is called when the click event is triggered.
```

Reply:
These are great suggestions! Also added a closing `</button>` tag to this other suggestion.